### PR TITLE
URLs to the output of report and query commands

### PR DESCRIPTION
**Pull request by smacker**
Fixes https://github.com/src-d/gemini/issues/9

### CODE REVIEW

**Comment by smacker** [comment]
Yeah. The issue was about github only. But it's better to support other services too. Bitbucket, gitlab, do we have anything else? Maybe retrieval team can help with a list of services? If I have time next week I'll make more general solution here.

**Comment by dpordomingo** [comment]
You may ask @vmarkovtsev ; he demoed something like that for Apolo

**Comment by vmarkovtsev** [comment]
AFAIK we do not have a fixed list, the repos can be from any source. I short-circuited for now: https://github.com/src-d/apollo/blob/master/apollo/query.py#L82

**Comment by vmarkovtsev** [comment]
BTW, I would suggest using https://github.com/HubSpot/jinjava so that we can share our report templates.

**Comment by dpordomingo** [comment]
Thanks for sharing that link!
I agree on the benefit of shareing conclusions of our investigations and I agree on the benefit of using a templating system
But I'm not sure about using the same templates/code. Since we are in an early stage of the development, it could be an overhead trying to maintain shared code.

**Comment by vmarkovtsev** [comment]
Since Apollo's side is Python, I can quickly and painlessly adapt to the changes in templates on Gemini side. So you lead the way and I follow here.

**Comment by smacker** [comment]
I think it's a good idea to use jinja style templates if we start to generate html reports. For now, it's just console output, doesn't worth it.

About bitbucket. To generate url for bb we need commit hash. Right now we don't save it in DB.
Do we really want to change schema because of it?

**Comment by bzz** [comment]
Nice improvement!

@smacker sorry for confusion, but it seems that schema needs to be changed anyway even for github case to include a reference - otherwise, could you help me understand how do we point to https://github.com/src-d/engine/blob/v0.1.x/.dockerignore vs https://github.com/src-d/engine/blob/master/.dockerignore ?

I think it would make sense to address here the 2 most common cases: Github and Bitbucket at the same time, #9 was updated with details on URL format.

And if that is correct,  given ref information adding a BitBucket url support should not require much effort.

**Comment by smacker** [comment]
Yes. The only problem was to decide to we want to add commit in DB or not. I don't feel it's up to me :) I'll proceed with the change.

**Comment by bzz** [comment]
@smacker 

Hm, what do you mean by
>  add commit in DB or not

Sorry if I did not explain it well above, but we do *not want* to have commit in DB (yet).

**Comment by smacker** [comment]
@bzz I'm confused now. You said we need to implement bitbucket and better github support here. But we __do__ need commit hash in DB for it.

**Comment by bzz** [comment]
After more discussion, decided to store and use [reference hash](https://github.com/src-d/engine/blob/master/src/main/scala/tech/sourced/engine/Schema.scala#L31) for URLs construction.

**Comment by smacker** [comment]
@bzz 
Could you please confirm that schema changes are correct (repo, ref_hash, file in this order).

And that `commit_hash` is always the same as `ref_hash`.
From my understanding of https://github.com/src-d/engine/pull/200 it is true.
Apollo also uses this field: https://github.com/src-d/apollo/blob/master/apollo/bags.py#L47
But still, I'm not 100% sure because the source code of engine isn't that straightforward for me.

File: src/main/scala/tech/sourced/gemini/Gemini.scala [hunk 1]
```diff
@@ -142,7 +143,23 @@ class Gemini(session: SparkSession, log: Slf4jLogger, keyspace: String = Gemini.
...
+    "github.com" ->  "https://%s/blob/%s/%s",
+    "bitbucket.org" -> "https://%s/src/%s/%s",
+    "gitlab.com" -> "https://%s/blob/%s/%s"
+  )
+  val default = ("", "repo: %s ref_hash: %s file: %s")
+
+  def format(repo: String, ref_hash: String, file: String): String = {
+    services.find{case (h, _) => repo.startsWith(h)}
+      .getOrElse(default)._2.format(repo.replaceFirst("\\.git$", ""), ref_hash, file)
+  }
```
Comment (bzz) [inline src/main/scala/tech/sourced/gemini/Gemini.scala]: Probably a nitpick, but method chaining might assist readability and is kind of customary to have whitespace around `{` like in http://twitter.github.io/effectivescala/#Collections-Style

Also, outer `{}` can be skipped in this case as well

**Comment by bzz** [comment]
> And that commit_hash is always the same as ref_hash.
yes, I think after `.getFirstReferenceCommit()` that's the correct way.

**Comment by smacker** [comment]
@bzz do you mean to add test case about correct `ref_hash`? Maybe I should add one more for `repository` too?

**Comment by smacker** [comment]
readability is improved.
test cases added.

**Comment by smacker** [comment]
Because of no further discussion. Rebased & Merged.

**Pull request merged**